polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -273,25 +276,42 @@ class IOSharedNotebook(
       }
     }
 
-    def runCell(id: CellID): IO[Stream[IO, Result]] = withInterpreterLaunch(id) {
-      kernel =>
-        val buf = new WindowBuffer[Result](1000)
-        kernel.runCell(id).map {
-          results =>
-            results.evalTap {
-              // buffer the result and also broadcast to all clients
-              result => IO(buf.add(result)) *> outputMessages.publish1(CellResult(ShortString(path), id, result))
-            }.evalTap {
-              // if there are any UpdatingDataReprs, watch for their updates and broadcast
-              case v: ResultValue => watchUpdatingValues(v).through(outputMessages.publish).compile.drain.start.as(()) // TODO: is it wise to forget the fiber?
-              case _ => IO.unit
-            }.onFinalize {
-              // write the buffered results to the notebook
-              ref.update {
-                case (ver, nb) => ver -> nb.setResults(id, buf.toList)
-              }
+    override def runCell(id: CellID): IO[Stream[IO, Result]] = queueCell(id).flatten
+
+    def queueCell(id: CellID): IO[IO[Stream[IO, Result]]] = get.flatMap {
+      notebook =>
+        notebook.getCell(id).filterNot(_.language == "text").fold[IO[IO[Stream[IO, Result]]]](IO.pure(IO.pure(Stream.empty))) {
```

Comment:
I wonder whether there's anything we could do to consolidate the implementations of the `KernelAPI` methods in `PolyKernel` and `Subscriber`. I'm worried about it getting a little confusing when things are subtly different across the two. 

For example, here `queueCell` filters out whether the cell language is `text`. In `PolyKernel`, that filtering is done in `withInterpreter` instead. Also, the equivalent method in `Subscriber` is called `withInterpreterLaunch`.

Reply:
Agreed, it bothers me too. Will try to change `SharedNotebook` to do only the things that only it can do (like writing outputs to the notebook instance)